README.md
[![Travis-CI Build Status](https://travis-ci.org/terrytangyuan/lfda.svg?branch=master)](https://travis-ci.org/terrytangyuan/lfda)
[![Coverage Status](https://coveralls.io/repos/terrytangyuan/lfda/badge.svg?branch=master)](https://coveralls.io/r/terrytangyuan/lfda?branch=master)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/lfda)](https://cran.r-project.org/package=lfda)
[![Downloads from the RStudio CRAN mirror](https://cranlogs.r-pkg.org/badges/grand-total/lfda)](https://cran.r-project.org/package=lfda)
[![License](http://img.shields.io/:license-mit-blue.svg?style=flat)](http://badges.mit-license.org)
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01572/status.svg)](https://doi.org/10.21105/joss.01572)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3352038.svg)](https://doi.org/10.5281/zenodo.3352038)

# lfda
R package for performing and visualizing *Local Fisher Discriminant Analysis*, *Kernel Local Fisher Discriminant Analysis*, and *Semi-supervised Local Fisher Discriminant Analysis*. It's the first package with those methods implemented in native R language. It also provides visualization functions to easily visualize the dimension reduction results.

Introduction to the algorithms and their application can be found [here](https://gastrograph.com/resources/whitepapers/local-fisher-discriminant-analysis-on-beer-style-clustering.html) and [here](http://www.ms.k.u-tokyo.ac.jp/software.html#LFDA). These methods are widely applied in feature extraction, dimensionality reduction, clustering, classification, information retrieval, and computer vision problems. An introduction to the package is also available in Chinese [here](https://cosx.org/2015/08/a-brief-description-of-the-method-and-the-algorithm-of-the-lfda-package/).

Welcome any [feedback](https://github.com/terrytangyuan/lfda/issues) and [pull request](https://github.com/terrytangyuan/lfda/pulls).  

## Install the current release from CRAN:
```{R}
install.packages('lfda')
```

## Install the latest development version from github:
```{R}
devtools::install_github('terrytangyuan/lfda')
```

## Citation

Please call `citation("lfda")` in R to properly cite this software. A white paper is published on Journal of Open Source Software [here](http://joss.theoj.org/papers/10.21105/joss.01572). 

## Examples
### Local Fisher Discriminant Analysis(LFDA)
Suppose we want to reduce the dimensionality of the original data set (we are using `iris` data set here) to 3, then we can run the following:
```{R}
k <- iris[,-5] # this matrix contains all the predictors to be transformed
y <- iris[,5] # this should be a vector that represents different classes
r <- 3 # dimensionality of the resulting matrix

# run the model, note that two other kinds metrics we can use: 'weighted' and 'orthonormalized'
model <- lfda(k, y, r, metric = "plain") 

plot(model, y) # 3D visualization of the resulting transformed data set

predict(model, iris[,-5]) # transform new data set using predict

```
### Kernel Local Fisher Discriminant Analysis(KLFDA)
The main usage is the same except for an additional `kmatrixGauss` call to the original data set to perform a kernel trick: 
```{R}
k <- kmatrixGauss(iris[,-5])
y <- iris[,5]
r <- 3
model <- klfda(k, y, r, metric = "plain")

```
Note that the `predict` method for klfda is still under development. The `plot` method works the same way as in `lfda`.

### Semi-supervised Local Fisher Discriminant Analysis(SELF)
This algorithm requires one additional argument such as `beta` that represents the degree of semi-supervisedness. Let's assume we ignore 10% of the labels in `iris` data set:
```{R}
k <- iris[,-5]
y <- iris[,5]
r <- 3
model <- self(k, y, beta = 0.1, r = 3, metric = "plain")

```
The methods `predict` and `plot` work the same way as in `lfda`. 
### Integration with {ggplot2::autoplot}
`{ggplot2::autoplot}` has been integrated with this package. Now `{lfda}` can be plotted in 2D easily and beautifully using `{ggfortify}` package. Go to [this link](http://rpubs.com/sinhrks/plot_pca) and scroll down to the last section for an example. 

## Contribute & Code of Conduct

To contribute to this project, please take a look at the [Contributing Guidelines](https://github.com/terrytangyuan/lfda/blob/master/CONTRIBUTING.md) first. Please note that this project is released with a [Contributor Code of Conduct](https://github.com/terrytangyuan/lfda/blob/master/CONDUCT.md). By contributing to this project, you agree to abide by its terms.

## Contact

Contact the maintainer of this package:
Yuan Tang <terrytangyuan@gmail.com>


NEWS
lfda 1.1.3
---------------------

* Added accepted JOSS paper.
* Enhanced documentation, examples and tests.
* Improved code style.

lfda 1.1.2
---------------------

* White paper has been uploaded to arXiv [here](https://arxiv.org/abs/1612.09219). Use `citation("lfda")` for more information.
* A couple of small warnings and fixes.

lfda 1.1.1
---------------------

* `ggfortify` package has integrated with `lfda`. Now it's easy to visualize `lfda`, `klfda`, and `self` in 2D with the help of  `ggplot2` and `ggfortify`.
* Move rgl to be suggested package to avoid installation problems due to rgl on mac. 

lfda 1.1.0
---------------------

* Added predict method for lfda so that the resulting transforming matrix can be easily applied to new data. 
* Added Semi-supervised Local Fisher Discriminant Analysis. 

lfda 1.0.0
---------------------

* R package for performing and visualizing *Local Fisher Discriminant Analysis*, and *Kernel Local Fisher Discriminant Analysis*.


tests/testthat.R
library(testthat)
library(lfda)

test_check("lfda")


tests/testthat/test_lfda.R
context("Test lfda functionalities")

data(iris)

test_that("lfda works correctly on different metrics", {
  k <- iris[, -5]
  y <- iris[, 5]
  r <- 3
  result <- lfda(k, y, r, metric = "plain")
  expect_equal(dim(result$T), c(4, 3))
  expect_equal(dim(result$Z), c(150, 3))

  result <- lfda(k, y, r, metric = "weighted")
  expect_equal(dim(result$T), c(4, 3))
  expect_equal(dim(result$Z), c(150, 3))

  result <- lfda(k, y, r, metric = "orthonormalized")
  expect_equal(dim(result$T), c(4, 3))
  expect_equal(dim(result$Z), c(150, 3))

  expect_error(lfda(k, y, r, metric = "plain", knn = 10000))

  # case when r = d
  result <- lfda(k, y, r = 4, metric = "plain")
  expect_equal(dim(result$T), c(4, 4))
  expect_equal(dim(result$Z), c(150, 4))
})

test_that("S3 methods on produced lfda model have the correct behaviors", {
  k <- iris[, -5]
  y <- iris[, 5]
  r <- 3

  model <- lfda(k, y, r = 4, metric = "plain")
  expect_equal(dim(predict(model, iris[, -5])), c(150, 4))
  expect_error(predict(model, iris[, -5], "wrongType"))
  expect_error(predict(model))

  expect_equal(length(capture.output(print(model))), 20)
})


tests/testthat/test_self.R
context("self")

data(iris)

test_that("self works", {
  X <- iris[, -5]
  Y <- iris[, 5]
  r <- 3
  result <- self(X, Y, beta = 0.1, r = 3, metric = "plain")
  expect_equal(dim(result$T), c(4, 3))
  expect_equal(dim(result$Z), c(150, 3))

  suppressWarnings(self(X, Y, beta = 0.1, r = 3, metric = "weighted"))
  result <- self(X, Y, beta = 0.1, r = 3, metric = "orthonormalized")
  expect_equal(dim(result$T), c(4, 3))
  expect_equal(dim(result$Z), c(150, 3))

  # case when r = d
  result <- self(X, Y, beta = 0.1, r = 4, metric = "plain")
  expect_equal(dim(result$T), c(4, 4))
  expect_equal(dim(result$Z), c(150, 4))

  # case when one label only has 4 obs, which is less than the default minObsPerLabel = 5
  expect_error(capture.output(
    self(iris[c(1:4, 51:dim(iris)[1]), -5], iris[c(1:4, 51:dim(iris)[1]), 5],
      beta = 0.1, r = 3, metric = "plain"
    )
  ))

  model <- self(X, Y, beta = 0.1, r = 3, metric = "plain")
  predict(model, iris[, -5])
  expect_error(predict(model, iris[, -5], "wrongType"))
  expect_error(predict(model))
  capture.output(print(model))
})


R/lfda.R
#' Local Fisher Discriminant Analysis for
#' Supervised Dimensionality Reduction
#'
#' Performs local fisher discriminant analysis (LFDA) on the given data.
#'
#' LFDA is a method for linear dimensionality reduction that maximizes
#' between-class scatter and minimizes within-class scatter while at the
#' same time maintain the local structure of the data so that multimodal
#' data can be embedded appropriately. Its limitation is that it only
#' looks for linear boundaries between clusters. In this case, a non-linear
#' version called kernel LFDA will be used instead. Three metric types can
#' be used if needed.
#'
#' @import rARPACK
#'
#' @export  lfda
#'
#' @param x n x d matrix of original samples.
#'          n is the number of samples.
#' @param y length n vector of class labels
#' @param r dimensionality of reduced space (default: d)
#' @param metric type of metric in the embedding space (no default)
#'               'weighted'        --- weighted eigenvectors
#'               'orthonormalized' --- orthonormalized
#'               'plain'           --- raw eigenvectors
#' @param knn parameter used in local scaling method (default: 5)
#'
#' @return list of the LFDA results:
#' \item{T}{d x r transformation matrix (Z = x * T)}
#' \item{Z}{n x r matrix of dimensionality reduced samples}
#'
#' @keywords lfda local fisher discriminant transformation mahalanobis metric
#'
#' @author Yuan Tang
#'
#' @seealso See \code{\link{klfda}} for the kernelized variant of
#'          LFDA (Kernel LFDA).
#'
#' @references
#' Sugiyama, M (2007).
#' Dimensionality reduction of multimodal labeled data by
#' local Fisher discriminant analysis.
#' \emph{Journal of Machine Learning Research}, vol.\bold{8}, 1027--1061.
#'
#' Sugiyama, M (2006).
#' Local Fisher discriminant analysis for supervised dimensionality reduction.
#' In W. W. Cohen and A. Moore (Eds.), \emph{Proceedings of 23rd International
#' Conference on Machine Learning (ICML2006)}, 905--912.
#'
#' @import rARPACK
#'
#' @examples
#'
#' k <- iris[, -5]
#' y <- iris[, 5]
#' r <- 3
#' lfda(k, y, r, metric = "plain")
lfda <- function(x, y, r, metric = c("orthonormalized", "plain", "weighted"), knn = 5) {
  metric <- match.arg(metric) # the type of the transforming matrix (metric)
  x <- t(as.matrix(x)) # transpose of original samples
  y <- t(as.matrix(y)) # transpose of original class labels
  d <- nrow(x) # number of predictors
  n <- ncol(x) # number of samples
  if (is.null(r)) r <- d # if no dimension reduction requested, set r to d

  tSb <- mat.or.vec(d, d) # initialize between-class scatter matrix (to be maximized)
  tSw <- mat.or.vec(d, d) # initialize within-class scatter matrix (to be minimized)

  # compute the optimal scatter matrices in a classwise manner
  for (i in unique(as.vector(t(y)))) {
    Xc <- x[, y == i] # data for this class
    nc <- ncol(Xc)

    # determine local scaling for locality-preserving projection
    Xc2 <- t(as.matrix(colSums(Xc^2)))
    # calculate the distance, using a self-defined repmat function that's the same
    # as repmat() in Matlab
    distance2 <- repmat(Xc2, nc, 1) + repmat(t(Xc2), 1, nc) - 2 * t(Xc) %*% Xc

    # Get affinity matrix
    A <- getAffinityMatrix(distance2, knn, nc)

    Xc1 <- as.matrix(rowSums(Xc))
    G <- Xc %*% (repmat(as.matrix(colSums(A)), 1, d) * t(Xc)) - Xc %*% A %*% t(Xc)
    tSb <- tSb + (G / n) + Xc %*% t(Xc) * (1 - nc / n) + Xc1 %*% (t(Xc1) / n)
    tSw <- tSw + G / nc
  }

  X1 <- as.matrix(rowSums(x))
  tSb <- tSb - X1 %*% t(X1) / n - tSw

  tSb <- (tSb + t(tSb)) / 2 # final between-class cluster matrix
  tSw <- (tSw + t(tSw)) / 2 # final within-class cluster matrix

  # find generalized eigenvalues and normalized eigenvectors of the problem
  if (r == d) {
    # without dimensionality reduction
    eigTmp <- eigen(solve(tSw) %*% tSb) # eigenvectors here are normalized
  } else {
    # dimensionality reduction (select only the r largest eigenvalues of the problem)
    eigTmp <- suppressWarnings(rARPACK::eigs(A = solve(tSw) %*% tSb, k = r, which = "LM")) # r largest magnitude eigenvalues
  }
  eigVec <- Re(eigTmp$vectors) # the raw transforming matrix
  eigVal <- as.matrix(Re(eigTmp$values))

  # options to require a particular type of returned transform matrix
  # transforming matrix (do not change the "=" in the switch statement)
  Tr <- getMetricOfType(metric, eigVec, eigVal, d)

  Z <- t(t(Tr) %*% x) # transformed data
  out <- list("T" = Tr, "Z" = Z)
  class(out) <- "lfda"
  return(out)
}
#' LFDA Transformation/Prediction on New Data
#'
#' This function transforms a data set, usually a testing set, using the trained LFDA metric
#' @param object The result from lfda function, which contains a transformed data and a transforming
#'        matrix that can be used for transforming testing set
#' @param newdata The data to be transformed
#' @param type The output type, in this case it defaults to "raw" since the output is a matrix
#' @param ... Additional arguments
#' @export
#' @method predict lfda
#' @return the transformed matrix
#' @author Yuan Tang
#'
#' @examples
#'
#' k <- iris[, -5]
#' y <- iris[, 5]
#' r <- 3
#' model <- lfda(k, y, r = 4, metric = "plain")
#' predict(model, iris[, -5])
predict.lfda <- function(object, newdata = NULL, type = "raw", ...) {
  if (is.null(newdata)) {
    stop("You must provide data to be used for transformation. ")
  }
  if (type != "raw") {
    stop('Types other than "raw" are currently unavailable. ')
  }
  if (is.data.frame(newdata)) newdata <- as.matrix(newdata)

  transformMatrix <- object$T

  result <- newdata %*% transformMatrix
  result
}
#' Print an lfda object
#'
#' Print an lfda object
#' @param x The result from lfda function, which contains a transformed data and a transforming
#' @param ... ignored
#' @export
#' @importFrom stats cov
#' @importFrom utils head
#' @method print lfda
print.lfda <- function(x, ...) {
  cat("Results for Local Fisher Discriminant Analysis \n\n")
  cat("The trained transforming matrix is: \n")
  print(head(x$T))

  cat("\n\n The original dataset after applying this metric transformation is:  \n")
  print(head(x$Z))

  cat("\n")
  cat("Only partial output is shown above. Please see the model output for more details. \n")
  invisible(x)
}


R/klfda.R
#' Kernel Local Fisher Discriminant Analysis for
#' Supervised Dimensionality Reduction
#'
#' Performs kernel local fisher discriminant analysis on the given data,
#' which is the non-linear version of LFDA (see details \code{\link{lfda}}).
#'
#' @param k n x n kernel matrix. Result of the \code{\link{kmatrixGauss}} function.
#'          n is the number of samples
#' @param y n dimensional vector of class labels
#' @param r dimensionality of reduced space (default: d)
#' @param metric type of metric in the embedding space (default: 'weighted')
#'               'weighted'        --- weighted eigenvectors
#'               'orthonormalized' --- orthonormalized
#'               'plain'           --- raw eigenvectors
#' @param knn parameter used in local scaling method (default: 6)
#' @param reg regularization parameter (default: 0.001)
#'
#' @return list of the LFDA results:
#' \item{T}{d x r transformation matrix (Z = t(T) * X)}
#' \item{Z}{r x n matrix of dimensionality reduced samples}
#'
#' @keywords klfda local fisher discriminant transformation mahalanobis metric
#'
#' @aliases klfda
#'
#' @author Yuan Tang
#'
#' @seealso See \code{\link{lfda}} for the linear version.
#'
#' @import rARPACK
#'
#' @export klfda
#'
#' @references
#' Sugiyama, M (2007). - contain implementation
#' Dimensionality reduction of multimodal labeled data by
#' local Fisher discriminant analysis.
#' \emph{Journal of Machine Learning Research}, vol.\bold{8}, 1027--1061.
#'
#' Sugiyama, M (2006).
#' Local Fisher discriminant analysis for supervised dimensionality reduction.
#' In W. W. Cohen and A. Moore (Eds.), \emph{Proceedings of 23rd International
#' Conference on Machine Learning (ICML2006)}, 905--912.
#'
#' Original Matlab Implementation: http://www.ms.k.u-tokyo.ac.jp/software.html#LFDA
#'
#' @examples
#'
#' k <- kmatrixGauss(iris[, -5])
#' y <- iris[, 5]
#' r <- 3
#' klfda(k, y, r, metric = "plain")
klfda <- function(k, y, r, metric = c("weighted", "orthonormalized", "plain"),
                  knn = 6, reg = 0.001) {
  metric <- match.arg(metric) # the type of the transforming matrix (metric)
  y <- t(as.matrix(y)) # transpose of original class labels
  n <- nrow(k) # number of samples
  if (is.null(r)) r <- n # if no dimension reduction requested, set r to n

  tSb <- mat.or.vec(n, n) # initialize between-class scatter matrix (to be maximized)
  tSw <- mat.or.vec(n, n) # initialize within-class scatter matrix (to be minimized)

  # compute the optimal scatter matrices in a classwise manner
  for (i in unique(as.vector(t(y)))) {
    Kcc <- k[y == i, y == i] # data for this class
    Kc <- k[, y == i]
    nc <- nrow(Kcc)

    # Define classwise affinity matrix
    Kccdiag <- diag(Kcc) # diagonals of the class-specific data
    distance2 <- repmat(Kccdiag, 1, nc) + repmat(t(Kccdiag), nc, 1) - 2 * Kcc

    # Get affinity matrix
    A <- getAffinityMatrix(distance2, knn, nc)

    Kc1 <- as.matrix(rowSums(Kc))
    Z <- Kc %*% (repmat(as.matrix(colSums(A)), 1, n) * t(Kc)) - Kc %*% A %*% t(Kc)
    tSb <- tSb + (Z / n) + Kc %*% t(Kc) * (1 - nc / n) + Kc1 %*% (t(Kc1) / n)
    tSw <- tSw + Z / nc
  }

  K1 <- as.matrix(rowSums(k))
  tSb <- tSb - K1 %*% t(K1) / n - tSw

  tSb <- (tSb + t(tSb)) / 2 # final between-class cluster matrix
  tSw <- (tSw + t(tSw)) / 2 # final within-class cluster matrix

  # find generalized eigenvalues and normalized eigenvectors of the problem
  eigTmp <- suppressWarnings(rARPACK::eigs(
    A = solve(tSw + reg * diag(1, nrow(tSw), ncol(tSw))) %*% tSb,
    k = r, which = "LM"
  )) # r largest magnitude eigenvalues
  eigVec <- Re(eigTmp$vectors) # the raw transforming matrix
  eigVal <- as.matrix(Re(eigTmp$values))

  # options to require a particular type of returned transform matrix
  # transforming matrix (do not change the "=" in the switch statement)
  Tr <- getMetricOfType(metric, eigVec, eigVal, n)

  Z <- t(t(Tr) %*% k) # transformed data
  out <- list("T" = Tr, "Z" = Z)
  class(out) <- "lfda"
  return(out)
}


R/helpers.R
#' Matlab-Syntaxed Repmat
#'
#' This function mimics the behavior and syntax of repmat() in Matlab
#' it generates a large matrix consisting of an N-by-M tiling copies of A
#'
#' @param A original matrix to be used as copies
#' @param N the number of rows of tiling copies of A
#' @param M the number of columns of tiling copies of A
#'
#' @return matrix consisting of an N-by-M tiling copies of A
#'
#' @export
#'
repmat <- function(A, N, M) {
  kronecker(matrix(1, N, M), A)
}
#' Negative One Half Matrix Power Operator
#'
#' This function defines operation for negative one half matrix
#' power operator
#'
#' @param x the matrix we want to operate on
#' @param n the exponent
#'
#' @return the matrix after negative one half power
#'
#' @export
#'
"%^%" <- function(x, n) {
  with(eigen(as.matrix(x)), vectors %*% (values^n * t(vectors)))
}
#' Get Affinity Matrix
#'
#' This function returns an affinity matrix within knn-nearest neighbors from the distance matrix.
#'
#' @param distance2 The distance matrix for each observation
#' @param knn The number of nearest neighbors
#' @param nc The number of observations for data in this class
#'
#' @export
#'
#' @return an affinity matrix - the larger the element in the matrix, the closer two data points are
getAffinityMatrix <- function(distance2, knn, nc) {
  sorted <- apply(distance2, 2, sort) # sort for each column by distance
  if (dim(sorted)[1] < knn + 1) {
    stop("knn is too large, please try to reduce it.")
  }
  kNNdist2 <- t(as.matrix(sorted[knn + 1, ])) # knn-th nearest neighbor
  sigma <- sqrt(kNNdist2)

  localscale <- t(sigma) %*% sigma
  # use only non-zero entries in localscale since this will be used in the denominator
  # to calculate the affinity matrix
  flag <- (localscale != 0)

  # define affinity matrix - the larger the element in the matrix, the closer two data points are
  A <- mat.or.vec(nc, nc)
  A[flag] <- exp(-distance2[flag] / localscale[flag])
  return(A)
}
#' Get Requested Type of Transforming Metric
#'
#' This function returns the requested type of transforming metric.
#'
#' @param metric The type of metric to be requested
#' @param eigVec The eigenvectors of the problem
#' @param eigVal The eigenvalues of the problem
#' @param total The number of total rows to be used for weighting denominator
#'
#' @return The transformation metric in requested type
#' @export
getMetricOfType <- function(metric, eigVec, eigVal, total) {
  return(switch(metric,
    # this weighting scheme is explained in section 3.3 in the first reference
    weighted = eigVec * repmat(t(sqrt(eigVal)), total, 1),
    orthonormalized = qr.Q(qr(eigVec)),
    plain = eigVec
  ))
}


R/kmatrixGauss.R
#' Gaussian Kernel Computation
#' (Particularly used in Kernel Local Fisher Discriminant Analysis)
#'
#' Gaussian kernel computation for klfda, which maps the original
#' data space to non-linear and higher dimensions.
#'
#' @param x n x d matrix of original samples.
#'          n is the number of samples.
#' @param sigma dimensionality of reduced space. (default: 1)
#'
#' @return K n x n kernel matrix.
#'           n is the number of samples.
#'
#' @keywords klfda kernel local fisher discriminant
#'           transformation mahalanobis metric
#'
#' @aliases kmatrixGauss
#'
#' @author Yuan Tang
#'
#' @seealso See \code{klfda} for the computation of
#'          kernel local fisher discriminant analysis
#'
#' @export kmatrixGauss
#'
#' @references
#' Sugiyama, M (2007).
#' Dimensionality reduction of multimodal labeled data by
#' local Fisher discriminant analysis.
#' \emph{Journal of Machine Learning Research}, vol.\bold{8}, 1027--1061.
#'
#' Sugiyama, M (2006).
#' Local Fisher discriminant analysis for supervised dimensionality reduction.
#' In W. W. Cohen and A. Moore (Eds.), \emph{Proceedings of 23rd International
#' Conference on Machine Learning (ICML2006)}, 905--912.
#'
#' https://shapeofdata.wordpress.com/2013/07/23/gaussian-kernels/
#'
#' @examples
#'
#' kmatrixGauss(iris[, -5])
kmatrixGauss <- function(x, sigma = 1) {
  x <- t(as.matrix(x))
  d <- nrow(x)
  n <- ncol(x)
  X2 <- t(as.matrix(colSums(x^2)))
  distance2 <- repmat(X2, n, 1) + repmat(t(X2), 1, n) - 2 * t(x) %*% x
  K <- exp(-distance2 / (2 * sigma^2))
  return(K)
}


R/lfda_visualization.R
#' Assigning Colors to A Vector
#'
#' This function assigns a color to each distinct value in the given vector.
#'
#' @param vec The vector where each distinct value will be assigned a color.
#'
#' @return The colors for each element in the given vector
#'
#' @import grDevices
#'
#' @export
#'
Cols <- function(vec) {
  cols <- grDevices::rainbow(length(unique(vec)))
  return(cols[as.numeric(as.factor(vec))])
}
#' 3D Visualization for LFDA/KLFDA Result
#'
#' This function plot 3 dimensions of the lfda/klfda result.
#'
#' @param x The lfda/klfda result.
#' @param labels A list of class labels used for lfda/klfda training.
#' @param cleanText A boolean value to specify whether to make the labels in the plot cleaner (default: FALSE)
#' @param ... Additional arguments
#'
#' @method plot lfda
#' @export
#'
#' @seealso See \code{\link{lfda}} and \code{\link{klfda}} for the metric learning method used for this visualization.
#'
plot.lfda <- function(x, labels, cleanText = FALSE, ...) {
  if (is.character(labels) | is.numeric(labels) | is.integer(labels)) {
    labels <- as.factor(labels)
  }
  if (!is.factor(labels)) {
    stop("labels need to be an object convertable to a factor.")
  }

  if (length(labels) != dim(x$Z)[1]) {
    stop("length of labels differs from the number of rows in x$Z.")
  }

  if (!is.logical(cleanText)) {
    stop("cleanText needs to be TRUE(T) or FALSE(F).")
  }

  if (dim(x$Z)[2] > 3) {
    x$Z <- x$Z[, 1:3]
    warning("dimensionality of the transformed data is larger than 3. Only the first 3 columns of the data will be used. ")
  } else if (dim(x$Z)[2] < 3) {
    stop("dimensionality of the transformed data must be larger than 3.")
  }

  transformedData <- as.data.frame(cbind(labels, x$Z))
  colnames(transformedData)[1] <- "Class"

  if (cleanText) {
    ## Show The Text of Each Style Only Once ##
    newData <- suppressWarnings( # known warnings
      plyr::ddply(transformedData, plyr::.(Class), function(y) {
        y[-1, "Class"] <- ""
        return(y)
      })
    )
    df <- sapply(newData[, "Class"], as.character)
    df[is.na(df)] <- "."
    newData[, "Class"] <- df
  } else {
    newData <- transformedData
  }

  # options(rgl.useNULL=TRUE) # deal with rgl in Travis

  ## Plot 3D Visualization of LFDA Result ##
  rgl::text3d(newData[, 2], newData[, 3], newData[, 4],
    col = Cols(transformedData$Class),
    main = "3D Visualization of Metric Transformed Data", size = 4,
    texts = newData$Class, font = 1, cex = 1.2, ...
  )
  rgl::axes3d()
}


R/metric_transformed_data_viz_3d.R
#' Assigning Colors to A Vector
#'
#' This function assigns a color to each distinct value in the given vector.
#'
#' @param vec The vector where each distinct value will be assigned a color.
#'
#' @return The colors for each element in the given vector
#'
#' @import grDevices
#'
#' @export
#'
Cols <- function(vec){
  cols <- grDevices::rainbow(length(unique(vec)))
  return(cols[as.numeric(as.factor(vec))])
}
#' 3D Visualization on Metric Transformed Data
#'
#' This function performs metric learning on the given dataset and runs
#' a 3D visualization on the transformed data.
#'
#' @param dat The original data.
#' @param cleanText A boolean value to specify whether the class labels need to be cleaned.(default:FALSE)
#' @param method The metric learning method used to transform the original data. It can be "lfda" or "klfda".
#'
#' @return list of the LFDA results:
#' \item{T}{d x r transformation matrix (Z = t(T) * X)}
#' \item{Z}{r x n matrix of dimensionality reduced samples}
#'
#' @seealso See \code{\link{lfda}} and \code{\link{klfda}} for the metric learning method used for this visualization.
#'
#' @import rgl plyr
#'
#' @export
#'
MetricTransformedViz <- function(dat, cleanText=FALSE, method=c("lfda", "klfda")){
  ## Apply Metric Learning ##
  k <- kmatrixGauss(x = dat[,-1]) # apply Gaussian Kernel
  y <- dat[,1]
  r <- 3 # dimensionality of reduced space
  method <- match.arg(method)
  ## if performing LFDA ##
  if(method=="lfda"){
    result <- lfda(dat[,-1],y,r,metric="plain")
    transformedData <- result$Z
    transformedData <- as.data.frame(cbind(dat[,1],transformedData))
    colnames(transformedData)[1] <- "Class"
  }
  ## if performing KLFDA ##
  else if(method=="klfda"){
    result <- klfda(k,y,r,metric="plain")
    transformedData <- as.data.frame(cbind(dat[,1],result$Z)) # metric transformed data
    colnames(transformedData)[1] <- "Class"
  }

  if(cleanText){
    ## Show The Text of Each Style Only Once ##
    newData <- suppressWarnings( # known warnings
      plyr::ddply(transformedData, plyr::.(Class), function(x){
        x[-1, "Class"] <- ""
        return(x)
      }))
    df <- sapply(newData[,"Class"], as.character)
    df[is.na(df)] <- "."
    newData[,"Class"] <- df
  } else {
    newData <- transformedData
  }


  ## 3D Visualization on Metric Transformed Dataset ##
  rgl::text3d(newData[,2],newData[,3], newData[,4], col=Cols(transformedData$Class),
              main="3D Visualization of Metric Transformed Data", size=4,
              texts=newData$Class, font=1, cex=1.2)
  rgl::axes3d()

  return(result)
}


R/self.R
#' Semi-Supervised Local Fisher Discriminant Analysis(SELF) for
#' Semi-Supervised Dimensionality Reduction
#'
#' Performs semi-supervised local fisher discriminant analysis (SELF) on the given data.
#' SELF is a linear semi-supervised dimensionality reduction method smoothly bridges supervised
#' LFDA and unsupervised principal component analysis, by which a natural regularization effect
#' can be obtained when only a small number of labeled samples are available.
#'
#' @import rARPACK
#'
#' @export  self
#'
#' @param X n x d matrix of original samples.
#'          n is the number of samples.
#' @param Y length n vector of class labels
#' @param beta degree of semi-supervisedness (0 <= beta <= 1; default is 0.5 )
#'        0: totally supervised (discard all unlabeled samples)
#'        1: totally unsupervised (discard all label information)
#' @param r dimensionality of reduced space (default: d)
#' @param metric type of metric in the embedding space (no default)
#'               'weighted'        --- weighted eigenvectors
#'               'orthonormalized' --- orthonormalized
#'               'plain'           --- raw eigenvectors
#' @param kNN parameter used in local scaling method (default: 5)
#' @param minObsPerLabel the minimum number observations required for each different label(default: 5)
#'
#' @return list of the SELF results:
#' \item{T}{d x r transformation matrix (Z = x * T)}
#' \item{Z}{n x r matrix of dimensionality reduced samples}
#'
#' @keywords lfda local fisher discriminant transformation mahalanobis metric semi-supervised
#'
#' @author Yuan Tang
#'
#' @seealso See \code{\link{lfda}} for LFDA and  \code{\link{klfda}} for the kernelized variant of
#'          LFDA (Kernel LFDA).
#'
#' @references
#' Sugiyama, Masashi, et al (2010).
#' Semi-supervised local Fisher discriminant analysis for dimensionality reduction.
#' \emph{Machine learning} 78.1-2: 35-61.
#'
#' Sugiyama, M (2007).
#' Dimensionality reduction of multimodal labeled data by
#' local Fisher discriminant analysis.
#' \emph{Journal of Machine Learning Research}, vol.\bold{8}, 1027--1061.
#'
#' Sugiyama, M (2006).
#' Local Fisher discriminant analysis for supervised dimensionality reduction.
#' In W. W. Cohen and A. Moore (Eds.), \emph{Proceedings of 23rd International
#' Conference on Machine Learning (ICML2006)}, 905--912.
#'
#' @import rARPACK
#'
#' @examples
#'
#' x <- iris[, -5]
#' y <- iris[, 5]
#' self(x, y, beta = 0.1, r = 3, metric = "plain")
self <- function(X, Y, beta = 0.5, r, metric = c("orthonormalized", "plain", "weighted"), kNN = 5, minObsPerLabel = 5) {
  if (any(table(Y) < minObsPerLabel)) {
    stop(cat(
      "Number of reviews per label is less than", minObsPerLabel, "\n",
      "the label(s):", unique(Y[which(table(Y) <= minObsPerLabel)]), "is/are the problem(s)!"
    ))
  }

  X <- t(as.matrix(X))
  Y <- t(as.matrix(Y))

  d <- nrow(X) # number of predictors
  n <- ncol(X) # number of samples
  if (is.null(r)) r <- d

  tSb <- mat.or.vec(d, d)
  tSw <- mat.or.vec(d, d)

  flag_label <- !is.na(Y)
  nlabel <- sum(flag_label)
  X2 <- t(as.matrix(colSums(X^2)))
  disttmp <- (repmat(X2[, flag_label], 1, nlabel) + repmat(t(X2), 1, nlabel) - 2 * t(X) %*% X[, flag_label]) # modified
  dist2 <- disttmp + abs(min(disttmp)) # modified
  A <- getAffinityMatrix(dist2, kNN, nlabel)

  Wlb <- mat.or.vec(nlabel, nlabel)
  Wlw <- mat.or.vec(nlabel, nlabel)
  Ylabel <- Y[flag_label]

  for (class in Ylabel) {
    flag_class <- (Ylabel == class)
    nclass <- sum(flag_class)
    if (nclass != 0) {
      tmp <- flag_class * 1
      tmp2 <- (!flag_class) * 1
      Wlb <- Wlb + (A * (1 / nlabel - 1 / nclass)) * as.numeric(t(tmp) %*% tmp) + as.numeric(t(tmp) %*% tmp2) / nlabel
      Wlw <- Wlw + (A / nclass) * as.numeric(t(tmp) %*% tmp)
    }
  }

  Slb <- X[, flag_label] %*% (diag(t(as.matrix(colSums(Wlb)))) - Wlb) %*% t(X[, flag_label])
  Slw <- X[, flag_label] %*% (diag(t(as.matrix(colSums(Wlw)))) - Wlw) %*% t(X[, flag_label])

  Srlb <- (1 - beta) * Slb + beta * cov(t(X)) / nrow(X)
  Srlw <- (1 - beta) * Slw + beta * diag(d)

  Srlb <- (Srlb + t(Srlb)) / 2
  Srlw <- (Srlw + t(Srlw)) / 2

  if (r == d) {
    eigTmp <- eigen(solve(Srlw) %*% Srlb)
  } else {
    eigTmp <- suppressWarnings(rARPACK::eigs(A = solve(Srlw) %*% Srlb, k = r, which = "LM"))
  }
  eigVec <- Re(eigTmp$vectors)
  eigVal <- as.matrix(Re(eigTmp$values))

  Tr <- getMetricOfType(metric, eigVec, eigVal, d)

  Z <- t(t(Tr) %*% X) # transformed data
  out <- list("T" = Tr, "Z" = Z)
  class(out) <- "lfda"
  return(out)
}
